keanu-project/src/main/java/io/improbable/keanu/network/LambdaSection.java

```diff
@@ -64,45 +89,63 @@ public static LambdaSection getUpstreamLambdaSection(Vertex<?> aVertex, boolean
      * includeNonProbabilistic is false.
      */
-    public static LambdaSection getDownstreamLambdaSection(Vertex<?> aVertex, boolean includeNonProbabilistic) {
+    public static LambdaSection getDownstreamLambdaSectionForCollection(List<Vertex> vertices, boolean includeNonProbabilistic) {
 
         Predicate<Vertex> shouldAdd = includeNonProbabilistic ? ADD_ALL : PROBABILISTIC_OR_OBSERVED_ONLY;
 
-        Set<Vertex> downstreamVertices = getVertices(
-            aVertex,
+        Set<Vertex> upstreamVertices = getVertices(
+            vertices,
             Vertex::getChildren,
             shouldAdd
         );
 
-        return new LambdaSection(downstreamVertices);
+        return new LambdaSection(upstreamVertices);
+    }
+
+    public static Set<Vertex> getVertices(Vertex vertex, Function<Vertex, Collection<Vertex>> nextVertices,
+                                          Predicate<Vertex> shouldAdd) {
+        return getVertices(Collections.singletonList(vertex), nextVertices, shouldAdd);
     }
 
     /**
-     * @param vertex       Vertex to start propagation from
+     * @param vertices     vertices to start propagation from
      * @param nextVertices The next vertices to move to given a current vertex. E.g getChildren for downstream or
      *                     getParents for upstream.
      * @param shouldAdd    true when a given vertex should be included in the result, false otherwise
      * @return A Set of vertices that are in the direction implied by nextVertices and filtered by shouldAdd
      */
-    public static Set<Vertex> getVertices(Vertex vertex,
-                                          Function<Vertex, Collection<Vertex>> nextVertices,
-                                          Predicate<Vertex> shouldAdd) {
+    public static Set<Vertex> getVertices(List<Vertex> vertices,
+                                        Function<Vertex, Collection<Vertex>> nextVertices,
+                                        Predicate<Vertex> shouldAdd) {
+
+        List<Vertex> verticesCopy = new ArrayList<>(vertices);
+        Vertex firstVertex = verticesCopy.remove(0);
+
+        Set<Vertex> queued = new HashSet<>(verticesCopy);
+        Set<Vertex> visited = new HashSet<>();
+        Deque<Vertex> stack = new ArrayDeque<>(verticesCopy);
+
+        Collection<Vertex> initialNext = nextVertices.apply(firstVertex);
+        queued.addAll(initialNext);
+        stack.addAll(initialNext);
 
-        Collection<Vertex> initialNext = nextVertices.apply(vertex);
-        Set<Vertex> queued = new HashSet<>(initialNext);
-        Deque<Vertex> stack = new ArrayDeque<>(initialNext);
         Set<Vertex> result = new HashSet<>();
-        result.add(vertex);
+        result.add(firstVertex);
+        result.addAll(verticesCopy);
 
         while (!stack.isEmpty()) {
             Vertex visiting = stack.pop();
 
+            if (visited.contains(visiting)) {
```

Comment:
Talked about this offline. Also, should add another test to LambdaSectionEfficiencyTest.